File: compressor.py
import math

class UnknownCompressor:
    def __init__(self, sample_rate=44100.0):
        # --- System Setup ---
        self.sample_rate = float(sample_rate)
        
        # --- Parameters (Knobs) ---
        # Corresponds to *(int *)(lVar4 + 0x9e54)
        # 0 = RMS, 2 = Peak
        self.mode = 2 
        
        # Corresponds to *(float *)(lVar4 + 0x9ee4)
        self.threshold = 1.0  # Linear scale (1.0 = 0dB)
        
        # Ratio is likely calculated externally and passed as a slope factor
        # We will simulate the ratio behavior here. 
        # 4.0 ratio means 1/4 output for input. Slope = 1.0 - (1.0/Ratio)
        self.ratio = 4.0 
        
        # Time constants (in seconds)
        # Corresponds to *(float *)(param_1 + 0xb0)
        self.attack_time = 0.002 
        self.release_time = 0.100
        
        # --- Internal State (History) ---
        # Corresponds to *(float *)(param_1 + 0xa8) -> RMS State
        self.rms_state = 0.0
        
        # Corresponds to *(float *)(param_1 + 0xc4) -> Peak State
        self.peak_state = 0.0
        
        # Corresponds to *(float *)(param_1 + 0xe4) -> Smoothed Gain State
        self.gain_state = 1.0
        
        # Constants for avoiding denormals/div-by-zero
        self.EPSILON = 9.536743e-07

    def _calc_coeff(self, time_seconds):
        """
        Approximates the coefficient calculation logic seen in:
        uVar14 = FUN_18103fd60(); 
        *(float *)(param_1 + 0xb8) = 1.0 - (float)uVar14;
        Standard 1-pole coeff: 1 - exp(-1 / (time * sample_rate))
        """
        if time_seconds <= 0.0:
            return 1.0
        # The decompiled function FUN_18103fd60 is a pow/exp approximation.
        # We use the standard math equivalent.
        return 1.0 - math.exp(-1.0 / (time_seconds * self.sample_rate))

    def process_block(self, left_channel, right_channel):
        """
        Functionally equivalent to: void unknown(longlong param_1, longlong *param_2)
        
        :param left_channel: List or numpy array of floats (Audio L)
        :param right_channel: List or numpy array of floats (Audio R)
        :return: (processed_left, processed_right)
        """
        
        # 1. Update Coefficients
        # In C, this checked if inputs changed. Here we calc every block for simplicity.
        attack_coeff = self._calc_coeff(self.attack_time)   # param_1 + 0xb8
        release_coeff = self._calc_coeff(self.release_time) # param_1 + 0xbc
        
        # RMS averaging coefficient (approximate constant for RMS mode)
        rms_coeff = self._calc_coeff(0.050) # param_1 + 0xa4 (Guessed value based on logic)
        
        # Gain smoothing coefficient (Ballistics)
        # The logic around uVar12/uVar13 suggests a smoothing filter for the gain reduction itself
        gain_smooth_coeff = self._calc_coeff(0.005) 

        # Calculate Slope for the gain computer: slope = 1 - (1/Ratio)
        # This matches the FUN_18103fd60 pow() logic in the gain reduction phase
        slope = 1.0 - (1.0 / self.ratio) if self.ratio > 1.0 else 0.0

        output_l = [0.0] * len(left_channel)
        output_r = [0.0] * len(right_channel)

        # 2. Processing Loop (do-while)
        for i in range(len(left_channel)):
            # Load Samples
            samp_l = left_channel[i] # fVar18
            samp_r = right_channel[i] # fVar19
            
            # --- DETECTOR STAGE ---
            current_env = 0.0 # fVar20
            
            if self.mode == 0: # RMS Mode (iVar2 == 0)
                # Calculate Power: (L^2 + R^2) * 0.5
                power = (samp_l * samp_l + samp_r * samp_r) * 0.5
                
                if power <= self.EPSILON:
                    power = self.EPSILON
                
                # Filter: state += (input - state) * coeff
                # Corresponds to: fVar20 = (fVar18 / fVar20 - fVar20) ... (The C math is optimized/weird here, standardizing)
                # Standard one-pole smoothing:
                self.rms_state += (power - self.rms_state) * rms_coeff
                current_env = math.sqrt(max(0, self.rms_state)) # RMS is Root Mean Square
                
            else: # Peak Mode (iVar2 == 2)
                abs_l = abs(samp_l)
                abs_r = abs(samp_r)
                
                # Take the max of absolute values
                input_level = abs_l if abs_l > abs_r else abs_r # *pfVar6
                
                # Attack / Release Branching
                # if input > current_state: coeff = attack, else: coeff = release
                coeff = attack_coeff if input_level > self.peak_state else release_coeff
                
                # Filter: fVar20 = (*pfVar6 - fVar20) * fVar18 + fVar20;
                self.peak_state += (input_level - self.peak_state) * coeff
                current_env = self.peak_state

            # --- GAIN COMPUTER STAGE ---
            # fVar1 is Threshold
            target_gain_linear = 1.0 # fVar18 (initialized to 1.0)
            
            if self.threshold < current_env:
                # Calculate Gain Reduction
                # C code uses: fVar18 = (float)uVar17 * (float)uVar15;
                # This represents the transfer function: Gain = (Threshold / Input) ^ Slope
                
                # Prevent div by zero
                safe_env = current_env if current_env > self.EPSILON else self.EPSILON
                ratio_calc = self.threshold / safe_env
                
                # The heavy FUN_18103fd60 call is pow(ratio_calc, slope)
                target_gain_linear = math.pow(ratio_calc, slope)

            # --- BALLISTICS (Smoothing the Gain) ---
            # fVar19 = *(float *)(param_1 + 0xe4);
            # fVar19 = (fVar18 - fVar19) * (1.0 - fVar11) + fVar19;
            self.gain_state += (target_gain_linear - self.gain_state) * gain_smooth_coeff
            
            applied_gain = self.gain_state # fVar19

            # --- APPLY GAIN ---
            # *(float *)(*param_2 + uVar16) = fVar19 * ...
            output_l[i] = samp_l * applied_gain
            output_r[i] = samp_r * applied_gain
            
            # --- METERING (Simplified) ---
            # The C code checks flags (0x9fb3, 0xa043) and updates pointers for UI meters.
            # We omit the raw pointer logic, but functionally, we would store max values here.
            # self.max_reduction = min(self.max_reduction, applied_gain)

        return output_l, output_r

# --- Example Usage ---
if __name__ == "__main__":
    # Initialize
    comp = UnknownCompressor(sample_rate=44100)
    
    # Configure (Matches C param_1 setup)
    comp.threshold = 0.5   # -6dB
    comp.ratio = 4.0       # 4:1
    comp.mode = 2          # Peak Mode
    comp.attack_time = 0.01
    comp.release_time = 0.1

    # Create dummy audio (Sine wave)
    data_len = 100
    input_l = [math.sin(x * 0.1) for x in range(data_len)]
    input_r = [math.cos(x * 0.1) for x in range(data_len)]

    # Process
    out_l, out_r = comp.process_block(input_l, input_r)

    print(f"Input [0]: {input_l[0]:.4f}, Output [0]: {out_l[0]:.4f}")
    print(f"Input [50]: {input_l[50]:.4f}, Output [50]: {out_l[50]:.4f}")

File: compressor.c

void compressor(longlong param_1, longlong *param_2)

{
    float fVar1;
    int iVar2;
    int iVar3;
    longlong lVar4;
    longlong lVar5;
    float *pfVar6;
    float *pfVar7;
    uint uVar8;
    longlong *plVar9;
    float fVar10;
    float fVar11;
    ulonglong uVar12;
    ulonglong uVar13;
    ulonglong uVar14;
    ulonglong uVar15;
    ulonglong uVar16;
    ulonglong uVar17;
    float fVar18;
    float fVar19;
    float fVar20;
    float local_res8[2];
    float local_res10[2];

    lVar4 = *(longlong *)(param_1 + 0x90);
    iVar2 = *(int *)(lVar4 + 0x9e54);
    fVar1 = *(float *)(lVar4 + 0x9ee4);
    (**(code **)(*(longlong *)(lVar4 + 0xa098) + 0x60))((longlong *)(lVar4 + 0xa098), *(undefined4 *)(lVar4 + 0x9ea8));
    local_res8[0] = 1.0 / *(float *)(param_1 + 0x98);
    uVar12 = FUN_18103fd60();
    uVar13 = FUN_18103fd60();
    if (iVar2 == 0)
    {
        *(undefined4 *)(param_1 + 0xc4) = *(undefined4 *)(param_1 + 0xc0);
    }
    else
    {
        *(undefined4 *)(param_1 + 0xa8) = 0x3ba3d70a;
        if (iVar2 == 2)
        {
            fVar10 = 0.0;
        }
        else
        {
            fVar10 = 0.002;
        }
        if (9.536743e-07 < ABS(*(float *)(param_1 + 0xb0) - fVar10))
        {
            *(float *)(param_1 + 0xb0) = fVar10;
            if (fVar10 <= 0.0)
            {
                *(undefined4 *)(param_1 + 0xb8) = 0x3f800000;
            }
            else
            {
                uVar14 = FUN_18103fd60();
                *(float *)(param_1 + 0xb8) = 1.0 - (float)uVar14;
            }
        }
    }
    lVar4 = *(longlong *)(param_1 + 0x90);
    uVar15 = FUN_18103fd60();
    lVar5 = param_2[5];
    uVar16 = FUN_18103fd60();
    fVar10 = local_res8[0];
    uVar14 = 0;
    fVar18 = (float)uVar16;
    *(undefined4 *)(param_1 + 0xe8) = 0;
    fVar19 = 0.0;
    if (*(char *)(param_1 + 0xfc) == '\0')
    {
        fVar20 = *(float *)(param_1 + 0xf4);
        if ((fVar18 - fVar20 <= -1e-15) || (1e-15 <= fVar18 - fVar20))
        {
            *(float *)(param_1 + 0xf0) = fVar20;
            fVar19 = 1.0 / (float)(int)lVar5;
            *(float *)(param_1 + 0xf8) = fVar20;
        }
        else
        {
            *(float *)(param_1 + 0xf0) = fVar18;
            fVar19 = 0.0;
            *(float *)(param_1 + 0xf8) = fVar18;
        }
    }
    else
    {
        *(float *)(param_1 + 0xf8) = fVar18;
        *(float *)(param_1 + 0xf0) = fVar18;
        *(undefined1 *)(param_1 + 0xfc) = 0;
    }
    *(float *)(param_1 + 0xf4) = fVar18;
    *(float *)(param_1 + 0xec) = fVar19;
    iVar3 = *(int *)(lVar4 + 0xa118);
    plVar9 = param_2;
    if (-1 < iVar3)
    {
        plVar9 = (longlong *)(param_2[3] + (longlong)iVar3 * 0x18);
    }
    uVar16 = uVar14;
    if (0 < (int)param_2[5])
    {
        do
        {
            lVar4 = *plVar9;
            fVar18 = *(float *)(param_1 + 0xe8) + 1.0;
            *(float *)(param_1 + 0xe8) = fVar18;
            fVar18 = fVar18 * *(float *)(param_1 + 0xec);
            *(float *)(param_1 + 0xf8) =
                (1.0 - fVar18) * *(float *)(param_1 + 0xf0) + fVar18 * *(float *)(param_1 + 0xf4);
            fVar18 = *(float *)(uVar16 + lVar4);
            fVar19 = *(float *)(uVar16 + plVar9[1]);
            if (iVar2 == 0)
            {
                fVar20 = *(float *)(param_1 + 0xa8);
                fVar18 = (fVar19 * fVar19 + fVar18 * fVar18) * 0.5;
                if (fVar18 <= 9.536743e-07)
                {
                    fVar18 = 9.536743e-07;
                }
                fVar20 = (fVar18 / fVar20 - fVar20) * *(float *)(param_1 + 0xa4) + fVar20;
                *(float *)(param_1 + 0xa8) = fVar20;
            }
            else
            {
                local_res8[0] = ABS(fVar18);
                fVar20 = *(float *)(param_1 + 0xc4);
                pfVar7 = local_res8;
                if (ABS(fVar18) <= ABS(fVar19))
                {
                    pfVar7 = local_res10;
                }
                local_res10[0] = ABS(fVar19);
                pfVar6 = (float *)(param_1 + 0xc0);
                if (*(float *)(param_1 + 0xc0) < *pfVar7)
                {
                    pfVar6 = pfVar7;
                }
                if (*pfVar6 <= fVar20)
                {
                    fVar18 = *(float *)(param_1 + 0xbc);
                }
                else
                {
                    fVar18 = *(float *)(param_1 + 0xb8);
                }
                fVar20 = (*pfVar6 - fVar20) * fVar18 + fVar20;
                *(float *)(param_1 + 0xc4) = fVar20;
            }
            fVar18 = 1.0;
            if (fVar1 < fVar20)
            {
                uVar17 = FUN_18103fd60();
                fVar18 = (float)uVar17 * (float)uVar15;
            }
            fVar19 = *(float *)(param_1 + 0xe4);
            fVar11 = (float)uVar13;
            if (fVar18 < fVar19)
            {
                fVar11 = (float)uVar12;
            }
            fVar19 = (fVar18 - fVar19) * (1.0 - fVar11) + fVar19;
            *(float *)(param_1 + 0xe4) = fVar19;
            fVar19 = fVar19 * *(float *)(param_1 + 0xf8);
            *(float *)(*param_2 + uVar16) = fVar19 * *(float *)(*param_2 + uVar16);
            *(float *)(param_2[1] + uVar16) = fVar19 * *(float *)(param_2[1] + uVar16);
            lVar4 = *(longlong *)(param_1 + 0x90);
            fVar18 = fVar20;
            if (fVar20 <= *(float *)(lVar4 + 0x9fbc))
            {
                fVar19 = *(float *)(param_1 + 0x100);
                if (fVar19 < 0.5)
                {
                    fVar18 = *(float *)(lVar4 + 0x9fbc);
                }
            }
            else
            {
                fVar19 = 0.0;
            }
            *(float *)(param_1 + 0x100) = fVar19 + fVar10;
            fVar19 = *(float *)(param_1 + 0xe0);
            if (fVar18 <= fVar19)
            {
                fVar11 = *(float *)(param_1 + 0xd8);
            }
            else
            {
                fVar11 = *(float *)(param_1 + 0xd4);
            }
            fVar19 = (fVar18 - fVar19) * fVar11 + fVar19;
            *(float *)(param_1 + 0xe0) = fVar19;
            *(float *)(lVar4 + 0x9fbc) = fVar19;
            if (*(char *)(lVar4 + 0x9ffb) == '\0')
            {
                local_res8[0] = *(float *)(lVar4 + 0x9fb8);
                *(undefined1 *)(lVar4 + 0x9ffb) = 1;
                lVar4 = *(longlong *)(lVar4 + 0xa128);
                pfVar7 = *(float **)(lVar4 + 0x30);
                if (pfVar7 == *(float **)(lVar4 + 0x38))
                {
                    FUN_1803ded20((longlong *)(lVar4 + 0x28), (undefined8 *)pfVar7, local_res8);
                }
                else
                {
                    *pfVar7 = local_res8[0];
                    *(longlong *)(lVar4 + 0x30) = *(longlong *)(lVar4 + 0x30) + 4;
                }
            }
            lVar4 = *(longlong *)(param_1 + 0x90);
            *(float *)(lVar4 + 0x9f74) = fVar20;
            if (*(char *)(lVar4 + 0x9fb3) == '\0')
            {
                local_res8[0] = *(float *)(lVar4 + 0x9f70);
                *(undefined1 *)(lVar4 + 0x9fb3) = 1;
                lVar4 = *(longlong *)(lVar4 + 0xa128);
                pfVar7 = *(float **)(lVar4 + 0x30);
                if (pfVar7 == *(float **)(lVar4 + 0x38))
                {
                    FUN_1803ded20((longlong *)(lVar4 + 0x28), (undefined8 *)pfVar7, local_res8);
                }
                else
                {
                    *pfVar7 = local_res8[0];
                    *(longlong *)(lVar4 + 0x30) = *(longlong *)(lVar4 + 0x30) + 4;
                }
            }
            lVar4 = *(longlong *)(param_1 + 0x90);
            local_res8[0] = *(float *)(param_1 + 0xe4);
            *(float *)(lVar4 + 0xa004) = local_res8[0];
            if (*(char *)(lVar4 + 0xa043) == '\0')
            {
                local_res8[0] = *(float *)(lVar4 + 0xa000);
                *(undefined1 *)(lVar4 + 0xa043) = 1;
                lVar4 = *(longlong *)(lVar4 + 0xa128);
                pfVar7 = *(float **)(lVar4 + 0x30);
                if (pfVar7 == *(float **)(lVar4 + 0x38))
                {
                    FUN_1803ded20((longlong *)(lVar4 + 0x28), (undefined8 *)pfVar7, local_res8);
                }
                else
                {
                    *pfVar7 = local_res8[0];
                    *(longlong *)(lVar4 + 0x30) = *(longlong *)(lVar4 + 0x30) + 4;
                }
            }
            uVar8 = (int)uVar14 + 1;
            uVar14 = (ulonglong)uVar8;
            uVar16 = uVar16 + 4;
        } while ((int)uVar8 < (int)param_2[5]);
    }
    return;
}
